### PR TITLE
Multi policies ipamless ipblock

### PR DESCRIPTION
**Pull request by maiqueb**
**- What this PR does and why is it needed**

This PR adds support for multi-network policies for secondary networks whose peers are of IPBlock type.

**- Special notes for reviewers**

**- How to verify it**

Run the multi-net tests.

**- Description for the changelog**

Add support for multi-network policies for secondary networks whose peers are of IPBlock type.

### CODE REVIEW

**Comment by maiqueb** [comment]
/cc @npinaeva 

Don't bother looking yet, the production code implementation will change - I don't like the code I've added.

But still, this has an e2e test to assert the feature, and right now, I'm just pushing the simplest thing that works.

**Comment by maiqueb** [comment]
/cc @cathy-zhou 

For your awareness. I plan on re-writing the code I'm adding to make it more readable, but for now I just want you to be able to track this work.

I'll let you know once this is ready for review. 

Thanks in advance.

**Comment by maiqueb** [comment]
FWIW, the e2e test I've added to check multi-network policies for IPAM less networks using IPBlock peers [is working](https://github.com/ovn-org/ovn-kubernetes/actions/runs/4892272519/jobs/8734333847?pr=3574#step:10:1600).

File: go-controller/pkg/ovn/base_network_controller_policy.go [hunk 1]
```diff
@@ -520,7 +520,7 @@ func (bnc *BaseNetworkController) getNewLocalPolicyPorts(np *networkPolicy,
 
 		// Skip pods that will never be present in logicalPortCache,
 		// e.g. hostNetwork pods, overlay node pods, or completed pods
-		if !bnc.podExpectedInLogicalCache(pod) {
+		if !bnc.podExpectedInLogicalCache(pod) && bnc.doesNetworkRequireIPAM() {
```
Comment (npinaeva) [inline go-controller/pkg/ovn/base_network_controller_policy.go:523]: Let's update `podExpectedInLogicalCache` to return false only for pods that won't be found in the `logicalPortCache`

File: go-controller/pkg/ovn/base_network_controller_policy.go [hunk 2]
```diff
@@ -1111,12 +1117,26 @@ func (bnc *BaseNetworkController) addNetworkPolicy(policy *knet.NetworkPolicy) e
...
+	} else {
+		var (
+			err      error
+			nsUnlock func()
+		)
+		aclLogging, nsUnlock, err = bnc.loadNamespaceAclLoggingConfig(policy.Namespace)
+		if err != nil {
+			return fmt.Errorf("failed to get nsInfo for namespace %s", policy.Namespace)
+		}
+		nsUnlock()
```
Comment (npinaeva) [inline go-controller/pkg/ovn/base_network_controller_policy.go:1138]: maybe it makes sense to create a function like "getNamespaceLogLevel" to encapsulate this logic?

File: go-controller/pkg/ovn/base_network_controller_policy.go [hunk 3]
```diff
@@ -1042,6 +1042,12 @@ func (bnc *BaseNetworkController) setupGressPolicy(np *networkPolicy, gp *gressP
 		gp.addIPBlock(peer.IPBlock)
 		return nil, nil
 	}
+
+	if !bnc.doesNetworkRequireIPAM() {
```
Comment (npinaeva) [inline go-controller/pkg/ovn/base_network_controller_policy.go:1046]: Maybe checking this earlier, e.g. in `convertMultiNetPolicyToNetPolicy` would be better

**Review by npinaeva — COMMENTED** [review]
some initial suggestions

File: go-controller/pkg/ovn/base_network_controller_policy.go (hunk 3 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_policy.go:1046]: It's going to be hard to do that inside that function, since that has no clue about the `net-attach-def` object, which makes it hard for us to understand if a network has IPAM or not. 


I'll try to think of something though ... I can definitely fail sooner than what I'm doing now.

File: go-controller/pkg/ovn/base_network_controller_policy.go (hunk 1 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_policy.go]: Done.

The better term fix would be to break the assumption a logical switch without a subnet represents an host network.

File: go-controller/pkg/ovn/base_network_controller_secondary.go [hunk 4]
```diff
@@ -58,6 +58,9 @@ func (bsnc *BaseSecondaryNetworkController) AddSecondaryNetworkResourceCommon(ob
 		if !ok {
 			return fmt.Errorf("could not cast %T object to *kapi.Namespace", obj)
 		}
+		if !bsnc.doesNetworkRequireIPAM() {
+			return nil
+		}
```
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:63]: Some questions here:

- Don't we need to react on ACL logging configuration changes?
- What about multicast? Is there any specific reason we would not support enabling/disabling multicast wich is also a namespace level configuration?
- Why are you not doing something similar in the `UpdateSecondaryNetworkResourceCommon` and `DeleteSecondaryNetworkResourceCommon` methods?

I see a possibility to keep processing namespace events being more "quirurgical" with what we do/don't do within the `AddNamespaceForSecondaryNetwork` business.

File: go-controller/pkg/ovn/base_network_controller_policy.go (hunk 3 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_policy.go:1046]: I'll explore these options.

IMO the first one is the cleaner option.

File: go-controller/pkg/ovn/base_network_controller_secondary.go (hunk 4 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:63]: > Some questions here:
> 
> * Don't we need to react on ACL logging configuration changes?

Right, this was from when I didn't care about ACL logging. 

> * What about multicast? Is there any specific reason we would not support enabling/disabling multicast wich is also a namespace level configuration?

True.

> * Why are you not doing something similar in the `UpdateSecondaryNetworkResourceCommon` and `DeleteSecondaryNetworkResourceCommon` methods?

Right. No good reason other than my own clumsiness. 

> 
> I see a possibility to keep processing namespace events being more "quirurgical" with what we do/don't do within the `AddNamespaceForSecondaryNetwork` business.

File: go-controller/pkg/ovn/base_network_controller_policy.go [hunk 5]
```diff
@@ -1111,12 +1112,25 @@ func (bnc *BaseNetworkController) addNetworkPolicy(policy *knet.NetworkPolicy) e
...
+		nsUnlock()
+	} else {
+		// we haven't kept track of the namespace in the cache; we need to load
+		// it from the informer. The nsInfo is cached as a side-effect.
+		nsInfo, nsUnlock, err := bnc.loadNamespace(policy.Namespace)
+		if err != nil {
+			return fmt.Errorf("failed to get nsInfo for namespace %s", policy.Namespace)
+		}
+		aclLogging = nsInfo.aclLogging
+		nsUnlock()
```
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_policy.go:1132]: I am bit confused here. 

I guess that if you just used `getNamespaceLocked` it wouldn't work because you would get back `nil` from it as there is no cached namespace info. But we are still watching for namespace events. So I guess that when we process those events, an error is happening preventing the cached namespace info from being populated, and is being kept retried to no avail. All for nothing because we don't expect that ever to succeed? Which means that we are watching namespaces for nothing and that we are not reacting namespace updates.

So I  think we should just use `getNamespaceLocked` and fix whatever issue is preventing the cache from being populated as a result of processing namespace events. I think the issue is just taking care of not checking for pod IPs on `ensureNamespaceLockedForSecondaryNetwork` and wherever `addPodToNamespaceForSecondaryNetwork` is being called from (the latter is already being handled).

Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_policy.go:1132]: > I am bit confused here.
> 
> I guess that if you just used `getNamespaceLocked` it wouldn't work because you would get back `nil` from it as there is no cached namespace info. But we are still watching for namespace events. So I guess that when we process those events, an error is happening preventing the cached namespace info from being populated, and is being kept retried to no avail. All for nothing because we don't expect that ever to succeed? Which means that we are watching namespaces for nothing and that we are not reacting namespace updates.
> 
> So I think we should just use `getNamespaceLocked` and fix whatever issue is preventing the cache from being populated as a result of processing namespace events. I think the issue is just taking care of not checking for pod IPs on `ensureNamespaceLockedForSecondaryNetwork` and wherever `addPodToNamespaceForSecondaryNetwork` is being called from (the latter is already being handled).

It's not only the IPs; I really don't see a need for even creating the address set (since afaiu, it's only purpose is to hold IPs that won't be there).

I would prefer to create another function similar to `ensureNamespaceLockedForSecondaryNetwork` that doesn't care about IPs, and doesn't create the address set. 
@jcaamano

**Comment by maiqueb** [comment]
0_o waaaat.

I guess I broke the PR when I tried to push a rebase + new code separately :S

File: go-controller/pkg/ovn/base_network_controller_policy.go (hunk 5 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_policy.go]: Trying a different approach that I think meets our ideas half-way. 

I agree this is more readable, and simpler to reason about.
@jcaamano

File: go-controller/pkg/ovn/base_network_controller_secondary.go [hunk 6]
```diff
@@ -448,6 +466,17 @@ func (bsnc *BaseSecondaryNetworkController) addPodToNamespaceForSecondaryNetwork
 	return ops, nil
 }
 
+// addPodToNamespaceForSecondaryNetwork just load the namespace information.
+func (bsnc *BaseSecondaryNetworkController) addPodToNamespaceForIPAMLessSecondaryNetwork(ns string) error {
+	var err error
```
Comment (maiqueb) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:471]: nit: not needed

File: go-controller/pkg/ovn/base_network_controller_secondary.go [hunk 7]
```diff
@@ -260,7 +260,8 @@ func (bsnc *BaseSecondaryNetworkController) ensureLocalZonePodForSecondaryNetwor
 func (bsnc *BaseSecondaryNetworkController) ensureRemoteZonePodForSecondaryNetwork(pod *kapi.Pod, addPort bool) error {
 
 	if !bsnc.doesNetworkRequireIPAM() {
-		return nil
+		_, err := bsnc.addPodToNamespaceForSecondaryNetwork(pod.Namespace, nil)
+		return err
```
Comment (npinaeva) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:264]: I see why we need to watch namespaces, but I don't understand why we need to react to the pod events with nil ips? the cached namespace will be created and updated based on namespace events only, and that is all we need?

Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:264]: I had proposed introducing a dedicated method that doesn't care about IPs / address sets.

I had an offline call w/ @jcaamano where he asked me to re-use as much code as possible. 

Thus, I ended up re-using that method.

Comment (npinaeva) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:264]: Yeah I agree with the general approach, let me rephrase my question: what exactly will not work if we revert only this specific change in `ensureRemoteZonePodForSecondaryNetwork` and leave everything else as is?

Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_secondary.go]: ah! My bad, didn't understand your original intention.

Seems you're right, and there's no need to actually keep track of these IPs for remote zones.

For the local zones, it **is** needed: we need to keep track of the LSP names for the port groups.

Good catch! 
@npinaeva

**Comment by jcaamano** [comment]
I think there is some of warnings we might want to get rid of.

Not sure we even want GetPodIPsOfNetwork to return an error, given that IPs are not mandatory:
https://github.com/ovn-org/ovn-kubernetes/blob/0e395d5a831930fd9f5e91d3303bf6d636da6136/go-controller/pkg/ovn/base_network_controller_namespace.go#L362-L366

Here neither the comment nor the returned error are very precise:
https://github.com/ovn-org/ovn-kubernetes/blob/0e395d5a831930fd9f5e91d3303bf6d636da6136/go-controller/pkg/util/pod_annotation.go#L302-L316

Again, here: 
https://github.com/ovn-org/ovn-kubernetes/blob/0e395d5a831930fd9f5e91d3303bf6d636da6136/go-controller/pkg/util/pod_annotation.go#L264-L268

File: go-controller/pkg/ovn/base_network_controller_namespace.go [hunk 8]
```diff
@@ -357,6 +357,9 @@ func (bnc *BaseNetworkController) getAllNamespacePodAddresses(ns string) []net.I
 	} else {
 		ips = make([]net.IP, 0, len(existingPods))
 		for _, pod := range existingPods {
+			if !bnc.doesNetworkRequireIPAM() {
```
Comment (npinaeva) [inline go-controller/pkg/ovn/base_network_controller_namespace.go:360]: you can just return an empty list with this condition at the very beginning of this function right? I mean no need to iterate over pods?

Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_namespace.go]: Done.

File: go-controller/pkg/ovn/base_network_controller_secondary.go [hunk 9]
```diff
@@ -288,7 +300,7 @@ func (bsnc *BaseSecondaryNetworkController) addLogicalPortToNetworkForNAD(pod *k
 		return err
 	}
 
-	if bsnc.doesNetworkRequireIPAM() && util.IsMultiNetworkPoliciesSupportEnabled() {
+	if util.IsMultiNetworkPoliciesSupportEnabled() {
 		// Ensure the namespace/nsInfo exists
 		addOps, err := bsnc.addPodToNamespaceForSecondaryNetwork(pod.Namespace, podAnnotation.IPs)
```
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:305]: Can you remind me why we need this to be called here on IPAM-less networks? I don't see the reason now.

Comment (maiqueb) [reply go-controller/pkg/ovn/base_network_controller_secondary.go]: Right, this isn't needed. Good catch.

Done.

File: go-controller/pkg/ovn/base_network_controller_namespace.go (hunk 8 above)
Comment (npinaeva) [reply go-controller/pkg/ovn/base_network_controller_namespace.go]: hehe never underestimate proper rest! thanks for the fast fix

**Review by npinaeva — APPROVED** [review]
Very nice PR, thanks @maiqueb !

**Comment by maiqueb** [comment]
rebasing in [this forced push](https://github.com/ovn-org/ovn-kubernetes/compare/6092c49b7ec9124fb4f0a1d9b1947229c8d4d3bd..faf5d850fb779385419a31fec56ecebaba4ff855) since it appears the bot is a bit incoherent today.

Hope now it goes back to work.

**Review by jcaamano — APPROVED** [review]
